### PR TITLE
feat: support config stream_route upstream in service

### PR DESCRIPTION
**Pull request by lingsamuel**
### Description

Support `service_id` for stream_route, allows config upstream in service.

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Comment by monkeyDluffy6017** [comment]
Please make the ci pass

File: apisix/plugin.lua [hunk 1]
```diff
@@ -636,6 +639,47 @@ function _M.merge_service_route(service_conf, route_conf)
...
+
+    if not new_conf.value.upstream and service_conf.value.upstream then
+        new_conf.value.upstream = service_conf.value.upstream
+    end
+
+    if not new_conf.value.upstream_id and service_conf.value.upstream_id then
+        new_conf.value.upstream_id = service_conf.value.upstream_id
+    end
+
+    --core.log.info("merged stream_route conf : ", require("inspect")(new_conf))
```
Comment (shreemaan-abhishek) [inline apisix/plugin.lua:668]: remove this line

**Comment by lingsamuel** [comment]
> Please make the ci pass

it can't pass in master
image

File: apisix/http/service.lua [hunk 2]
```diff
@@ -61,7 +61,7 @@ function _M.init_worker()
         filter = filter,
     })
     if not services then
-        error("failed to create etcd instance for fetching upstream: " .. err)
+        error("failed to create etcd instance for fetching /services: " .. err)
```
[the remark this answers is not in the document]
Comment (lingsamuel) [reply apisix/http/service.lua:64]: other resources have similar format. it indicates the etcd key

File: apisix/init.lua [hunk 3]
```diff
@@ -1078,6 +1079,34 @@ function _M.stream_preread_phase()
...
+        api_ctx.service_name = service.value.name
+        api_ctx.matched_upstream = matched_route.value.upstream
+        if matched_route.value.upstream_id and not matched_route.value.upstream then
+            local upstream = apisix_upstream.get_by_id(matched_route.value.upstream_id)
+            if not upstream then
+                if is_http then
+                    return core.response.exit(502)
+                end
+
+                return ngx_exit(1)
```
[the remark this answers is not in the document]
Comment (lingsamuel) [reply apisix/init.lua:1105]: because no upstream available here

**Comment by monkeyDluffy6017** [comment]
Please make the ci pass

File: apisix/init.lua (hunk 3 above)
[the remark this answers is not in the document]
Comment (shreemaan-abhishek) [reply apisix/init.lua:1105]: but in two lines above, there is: `response.exit(502)`. Why is it not sufficient? Why do you further check for `is_http` and only then return 502?

**Comment by monkeyDluffy6017** [comment]
Please update the related doc: https://github.com/apache/apisix/blob/master/docs/en/latest/admin-api.md

File: apisix/init.lua [hunk 4]
```diff
@@ -1021,6 +1021,7 @@ function _M.stream_init_worker()
     plugin.init_worker()
     xrpc.init_worker()
     router.stream_init_worker()
+    require("apisix.http.service").init_worker()
```
Comment (monkeyDluffy6017) [inline apisix/init.lua:1024]: Can the `service.lua` be moved out of the `http` directory?

File: apisix/plugin.lua [hunk 5]
```diff
@@ -636,6 +639,46 @@ function _M.merge_service_route(service_conf, route_conf)
...
+    end
+
+    if not new_conf.value.upstream_id and service_conf.value.upstream_id then
+        new_conf.value.upstream_id = service_conf.value.upstream_id
+    end
+
+    return new_conf
+end
+
+function _M.merge_service_stream_route(service_conf, route_conf)
```
Comment (monkeyDluffy6017) [inline apisix/plugin.lua:671]: Two blanks are needed between functions

File: t/stream-node/sanity-with-service.t [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+        }
+    }
+--- request
+GET /t
+--- response_body
+passed
+
+
+
+=== TEST 6: set stream route(id: 1) with server port
```
Comment (monkeyDluffy6017) [inline t/stream-node/sanity-with-service.t:130]: Why need this test case?

File: t/stream-node/sanity-with-service.t [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+
+=== TEST 14: hit route
+--- stream_request eval
+mmm
+--- stream_response
+hello world
+
+
+
+=== TEST 15: reuse ctx and more
```
Comment (monkeyDluffy6017) [inline t/stream-node/sanity-with-service.t:325]: Why need this test case?

File: apisix/init.lua (hunk 3 above)
[the remark this answers is not in the document]
Comment (monkeyDluffy6017) [reply apisix/init.lua:1105]: @shreemaan-abhishek it's just copy from the code above
```
    local up_id = matched_route.value.upstream_id
    if up_id then
        local upstream = apisix_upstream.get_by_id(up_id)
        if not upstream then
            if is_http then
                return core.response.exit(502)
            end

            return ngx_exit(1)
        end

        api_ctx.matched_upstream = upstream
```

File: apisix/init.lua (hunk 4 above)
Comment (lingsamuel) [reply apisix/init.lua:1024]: no, obviously service is designed for http

**Review by monkeyDluffy6017 — APPROVED** [review]

**Review by shreemaan-abhishek — APPROVED** [review]

**Pull request merged**